File: Android/app/src/main/java/com/arcinteractive/spaces/data/model/SpaceMember.kt
package com.arcinteractive.spaces.data.model

data class SpaceMember(
    val id: String,
    val displayName: String,
    val emojiAvatar: String,
    val role: SpaceMemberRole,
    val status: String
)

enum class SpaceMemberRole(
    val label: String
) {
    Owner("Owner"),
    Admin("Admin"),
    Moderator("Moderator"),
    Member("Member"),
    Guest("Guest");

    val firestoreValue: String
        get() = name.lowercase()

    companion object {
        fun fromFirestoreValue(value: String): SpaceMemberRole? = entries.firstOrNull {
            it.firestoreValue == value.lowercase()
        }
    }
}

File: iOS/Spaces V1.6/Models/SpaceMember.swift
import Foundation

struct SpaceMember: Identifiable, Hashable {
    let id: String
    let displayName: String
    let emojiAvatar: String
    let role: SpaceMemberRole
    let status: String

    func hasPermission(_ permission: SpacePermission) -> Bool {
        role.capabilities.contains(permission)
    }
}

enum SpaceMemberRole: String, Codable, CaseIterable, Identifiable, Hashable {
    case owner = "Owner"
    case admin = "Admin"
    case moderator = "Moderator"
    case member = "Member"
    case guest = "Guest"

    var id: String { rawValue }

    var firestoreValue: String {
        rawValue.lowercased()
    }

    init?(firestoreValue: String) {
        switch firestoreValue.lowercased() {
        case "owner":
            self = .owner
        case "admin":
            self = .admin
        case "moderator":
            self = .moderator
        case "member":
            self = .member
        case "guest":
            self = .guest
        default:
            return nil
        }
    }
}
